### PR TITLE
Strange beacon branch

### PR DESCRIPTION
**Pull request by Unidentified-Biohazard**
## About The Pull Request

Adds a strange beacon to the game in Science, on a cool down can create a strange object (can be upgraded)
## Why It's Good For The Game

Strange objects are rarely used, hopefully this will change that.
## Changelog
:cl:
add: Added the strange beacon, which costs one scanning module and one bluespace crystal. Can summon strange objects on a cooldown.
code: Added the strange beacon circuit to the circuits code.
/:cl:

### CODE REVIEW

**Comment by Blueturbo47** [comment]
You don't have to create a new fork, just make a new branch.

File: beestation.dme [hunk 1]
```diff
@@ -709,6 +709,7 @@
 #include "code\game\machinery\spaceheater.dm"
 #include "code\game\machinery\stasis.dm"
 #include "code\game\machinery\status_display.dm"
+#include "code\game\machinery\Strangebeacon.dm"
```
Comment (utilbird) [inline beestation.dme]: Try keeping file names lowercase fir consistency

File: code/game/machinery/Strangebeacon.dm [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		docreate()
+		lastUse = world.time
+	else
+		to_chat(user, "<span class='notice'>The beacon is still recharging.</span>")
+
+/obj/machinery/strange_beacon/attackby(obj/item/G, mob/user, params)
+	if(G.tool_behaviour == TOOL_SCREWDRIVER)
+		deconstruct()
+
+/obj/machinery/strange_beacon/proc/docreate()
```
Comment (utilbird) [inline code/game/machinery/Strangebeacon.dm]: Do you need a seperate proc for this?

Comment (Archanial) [inline code/game/machinery/Strangebeacon.dm]:
```suggestion
/obj/machinery/strange_beacon/proc/do_create()
```
just to be consistent with rest of the procs

**Review by Archanial — COMMENTED** [review]
Also you should rename Strangebeacon.dm to lowercase strangebeacon just to be consistent with the rest of the files.

File: code/game/machinery/Strangebeacon.dm [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		playsound(get_turf(src), 'sound/weapons/flash.ogg', 25, TRUE)
+		do_sparks(5,FALSE, src)
+		docreate()
+		lastUse = world.time
+	else
+		to_chat(user, "<span class='notice'>The beacon is still recharging.</span>")
+
+/obj/machinery/strange_beacon/attackby(obj/item/G, mob/user, params)
+	if(G.tool_behaviour == TOOL_SCREWDRIVER)
+		deconstruct()
```
Comment (Archanial) [inline code/game/machinery/Strangebeacon.dm]: also shouldn't you do the ..() at the end here?
Anyway, screwdrivers typically open the machines so you can upgrade them. Deconstructing it right away will feel really weird.

File: code/game/machinery/Strangebeacon.dm [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	RefreshParts()
+
+/obj/machinery/strange_beacon/RefreshParts()
+	coolDown = 0
+	for(var/obj/item/stock_parts/scanning_module/S in component_parts)
+		if(S.rating)
+			coolDown += 3600/S.rating
+
+/obj/machinery/strange_beacon/interact(mob/user)
+	add_fingerprint(user)
```
Comment (utilbird) [inline code/game/machinery/Strangebeacon.dm]: finger prints are already added in the parent proc
```suggestion
	..()
```

File: code/game/machinery/Strangebeacon.dm (hunk 3 above)
Comment (Unidentified-Biohazard) [reply code/game/machinery/Strangebeacon.dm]: I am pretty sure that some machines do immediately deconstruct on a screwdriver, though I can't think of an example now.

File: code/game/machinery/Strangebeacon.dm (hunk 4 above)
Comment (Unidentified-Biohazard) [reply code/game/machinery/Strangebeacon.dm]: noted

File: beestation.dme (hunk 1 above)
Comment (Unidentified-Biohazard) [reply beestation.dme]: Done

File: code/game/machinery/Strangebeacon.dm (hunk 2 above)
Comment (Unidentified-Biohazard) [reply code/game/machinery/Strangebeacon.dm]: yeah that way it can be easily added onto.

Comment (Unidentified-Biohazard) [reply code/game/machinery/Strangebeacon.dm]: done

File: code/game/machinery/Strangebeacon.dm (hunk 3 above)
Comment (Unidentified-Biohazard) [reply code/game/machinery/Strangebeacon.dm]: eh I did it anyway

Comment (Unidentified-Biohazard) [reply code/game/machinery/Strangebeacon.dm]: also, why would I put ..() there?

File: code/game/machinery/strange_beacon.dm [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+/obj/machinery/strange_beacon/Initialize()
+	. = ..()
+	RefreshParts()
+
+/obj/machinery/strange_beacon/RefreshParts()
+	coolDown = 0
+	for(var/obj/item/stock_parts/scanning_module/S in component_parts)
+		if(S.rating)
+			coolDown += 3600/S.rating
```
Comment (Archanial) [inline code/game/machinery/strange_beacon.dm:23]:
```suggestion
	coolDown = initial(coolDown)
	for(var/obj/item/stock_parts/scanning_module/S in component_parts)
		if(S.rating)
			coolDown = coolDown/S.rating
```

File: code/game/machinery/strange_beacon.dm [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+		new /obj/item/relic(loc)
+		playsound(get_turf(src), 'sound/weapons/flash.ogg', 25, TRUE)
+		do_sparks(5,FALSE, src)
+		lastUse = world.time
+	else
+		to_chat(user, "<span class='notice'>The beacon is still recharging.</span>")
+
+/obj/machinery/strange_beacon/attackby(obj/item/G, mob/user, params)
+	if(G.tool_behaviour == TOOL_SCREWDRIVER)
+		if(panel_open == FALSE)
```
Comment (Archanial) [inline code/game/machinery/strange_beacon.dm]:
```suggestion
		if(!panel_open)
```

File: code/game/machinery/strange_beacon.dm [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+			panel_open = TRUE
+		else
+			panel_open = FALSE
+	if(G.tool_behaviour == TOOL_CROWBAR && panel_open)
+		default_deconstruction_crowbar(G)
```
Comment (Archanial) [inline code/game/machinery/strange_beacon.dm:47]:
```suggestion
		default_deconstruction_crowbar(G)
	..()
```
shouldn't this also be here?

File: code/game/machinery/strange_beacon.dm (hunk 5 above)
Comment (Unidentified-Biohazard) [reply code/game/machinery/strange_beacon.dm:23]: it doesn't really make a difference?

File: code/game/machinery/strange_beacon.dm (hunk 7 above)
Comment (Unidentified-Biohazard) [reply code/game/machinery/strange_beacon.dm:47]: Might I ask why?

File: code/game/machinery/strange_beacon.dm [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+		to_chat(user, "<span class='notice'>The beacon is still recharging.</span>")
+
+/obj/machinery/strange_beacon/attackby(obj/item/G, mob/user, params)
+	if(G.tool_behaviour == TOOL_SCREWDRIVER)
+		if(!panel_open)
+			panel_open = TRUE
+		else
+			panel_open = FALSE
+	if(G.tool_behaviour == TOOL_CROWBAR && panel_open)
+		default_deconstruction_crowbar(G)
```
Comment (park66665) [inline code/game/machinery/strange_beacon.dm:47]:
```suggestion
/obj/machinery/strange_beacon/attackby(obj/item/I, mob/user, params)
	if(default_deconstruction_screwdriver(user, initial(icon_state), initial(icon_state), I))
		return
	if(default_deconstruction_crowbar(I))
		return
	return ..()
```

`default_deconstruction_screwdriver` handles screwdriver checking, panel open & close. The same goes for `default_deconstruction_crowbar(I)`. It does crowbar checks and other essential stuff. Finally, parent proc calling at `attackby` is *very* recommended.

File: code/game/machinery/strange_beacon.dm [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	idle_power_usage = 25
+	active_power_usage = 500
+	obj_flags = CAN_BE_HIT
+	circuit = /obj/item/circuitboard/machine/strange_beacon
+	var/coolDown = 3600
+	var/lastUse
+
+/obj/machinery/strange_beacon/Initialize()
+	. = ..()
+	RefreshParts()
```
Comment (park66665) [inline code/game/machinery/strange_beacon.dm:17]:
```suggestion
	RefreshParts()
	lastUse = world.time
```

Without this, you can mass-produce strange objects by constantly rebuilding it.

File: code/game/machinery/strange_beacon.dm (hunk 5 above)
Comment (park66665) [inline code/game/machinery/strange_beacon.dm:23]:
```suggestion
	coolDown = initial(coolDown)
	for(var/obj/item/stock_parts/scanning_module/S in component_parts)
		coolDown -= 600*S.rating
```

**Review by park66665 — CHANGES_REQUESTED** [review]

**Comment by utilbird** [comment]
I can't imagine giving people the ability to constantly generate strange objects with little to no effort will have any positives effects, gameplay wise.

**Comment by utilbird** [comment]
Closing due to inactivity, feel free to re-open once this is being worked on again

**Pull request closed without merging**